invenio_pidstore/providers/base32.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015-2018 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Douglas Crockford Base32-URL encoder.
+
+This encoder/decoder:
+- uses Douglas Crockford Base32 encoding
+- allows for checksum
+- encodes the checksum using only characters in the base32 set
+  (only digits in fact)
+- produces string that are URI-friendly (no '=' or '/' for instance)
+
+This is based on:
+- https://github.com/datacite/base32-url
+- https://github.com/jbittel/base32-crockford
+"""
+import string
+
+import six
+
+# NO i, l, o or u
+ENCODING_CHARS = '0123456789abcdefghjkmnpqrstvwxyz'
+DECODING_CHARS = {c: i for i, c in enumerate(ENCODING_CHARS)}
+
+
+def encode(number, split_every=0, checksum=False):
+    """Encodes `number` to URI-friendly Douglas Crockford base32 string.
+
+    :param number: number to encode
+    :param split_every: if provided, insert '-' every `split_every` characters
+                        going from left to right
+    :param checksum: append modulo 97-10 (ISO 7064) checksum to string
+    :returns: A random Douglas Crockford base32 encoded string composed only
+              of valid URI characters.
+    """
+    assert isinstance(number, six.integer_types)
+
+    if number < 0:
+        raise ValueError("Invalid 'number'. 'number' must > 0.")
+
+    if split_every < 0:
+        raise ValueError("Invalid 'split_every'. 'split_every' must > 0.")
+
+    encoded = ''
+    original_number = number
+    while number > 0:
+        remainder = number % 32
+        number //= 32
+        encoded = ENCODING_CHARS[remainder] + encoded
+
+    if checksum:
+        # NOTE: 100 * original_number is used because datacite also uses it
+        computed_checksum = 97 - ((100 * original_number) % 97) + 1
+        encoded_checksum = "{:02d}".format(computed_checksum)
+        encoded += encoded_checksum
+
+    if split_every:
+        splits = [
+            encoded[i:i+split_every]
+            for i in range(0, len(encoded), split_every)
+        ]
+        encoded = '-'.join(splits)
+
+    return encoded
+
+
+def normalize(encoded):
+    """Returns normalized encoded string.
+
+    - string is lowercased
+    - '-' are removed
+    - I,i,l,L decodes to the digit 1
+    - O,o decodes to the digit 0
+
+    :param encoded: string to decode
+    :returns: normalized string.
+    """
+    table = (
```

Comment:
**Minor**: (I see there is a test for it) Removing `U` is not needed because its not in the encoding characters, right?

Reply:
Yes, if the decoder gets `IOU20D011ARS`, it will raise a `ValueError("'encoded' contains undecodable characters")` because of the `U`. But this isn't explicitly tested so I am adding a test for `U` because it is an edge case. (The `Ü` test was for unicode characters).